### PR TITLE
Implement a lint to replace manual bit rotations with rotate_left/rot…

### LINKED ISSUE #6861: Suggest replacing hand made integer rotations with rotate_right/rotate_left
I think Clippy should suggest to replace integer rotations like:

```rust
pub fn rol24a(x: u32) -> u32 {
    (x >> 8) | (x << 24)
}
```

With calls to rotation methods `rotate_right` or `rotate_left`:

```rust
pub fn rol24b(x: u32) -> u32 {
    x.rotate_right(8)
}
```

### PR DESCRIPTION
**Pull request by frp**
Fixes #6861 

r? @llogiq 

---

changelog: add [`manual_rotate`] lint

### CODE REVIEW

File: clippy_lints/src/manual_rotate.rs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    /// Calling the function better conveys the intent.
+    ///
+    /// ### Known issues
+    ///
+    /// Currently, the lint only catches shifts by constant amount.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// (x >> 8) | (x << 24)
+    /// ```
```
Comment (llogiq) [inline clippy_lints/src/manual_rotate.rs:29]: It fails to compile the doc test.

```suggestion
    /// ```no_run
    /// let x = 12345678_u32;
    /// (x >> 8) | (x << 24)
    /// ```
```

File: clippy_lints/src/manual_rotate.rs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    /// Use instead:
+    /// ```no_run
+    /// x.rotate_right(8)
+    /// ```
```
Comment (llogiq) [inline clippy_lints/src/manual_rotate.rs:33]: Same here.

**Review by llogiq — CHANGES_REQUESTED** [review]
The doc tests currently fail to compile. I'll gladly approve this PR once that's fixed.

File: clippy_lints/src/manual_rotate.rs (hunk 1 above)
Comment (frp) [reply clippy_lints/src/manual_rotate.rs]: Done.

File: clippy_lints/src/manual_rotate.rs (hunk 2 above)
Comment (frp) [reply clippy_lints/src/manual_rotate.rs]: Done

**Comment by llogiq** [comment]
Thank you!

@bors r+

**Comment by llogiq** [comment]
I cannot see how that error would be related. Might be transient.

@bors retry

**Comment by frp** [comment]
I can reproduce the same failure locally by checking out `master` and re-running the integration test. So it's unrelated to my change.

The last change on `master` is https://github.com/rust-lang/rust-clippy/pull/12965, and this test passed when it was merged (16h ago).

I _think_ this commit in cargo might be the cause: https://github.com/rust-lang/cargo/commit/28d213bae29ba7c8fe957d496bdd342acdc3ea8b - it adds the exact lines the integration test is unhappy about.

How do we proceed?

**Comment by llogiq** [comment]
I have pushed a fix, which was just merged to master. If you pull & rebase onto that, it should work.

**Comment by llogiq** [comment]
Thank you!

@bors r+

File: clippy_lints/src/manual_rotate.rs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    fn check_expr<'tcx>(&mut self, cx: &LateContext<'tcx>, expr: &Expr<'tcx>) {
+        if let ExprKind::Binary(op, l, r) = expr.kind
+            && let BinOpKind::Add | BinOpKind::BitOr = op.node
+            && let Some((l_shift_dir, l_amount, l_expr)) = parse_shift(cx, l)
+            && let Some((r_shift_dir, r_amount, r_expr)) = parse_shift(cx, r)
+        {
+            if l_shift_dir == r_shift_dir {
+                return;
+            }
+            if !clippy_utils::SpanlessEq::new(cx).eq_expr(l_expr, r_expr) {
```
Comment (Alexendoo) [inline clippy_lints/src/manual_rotate.rs:87]: I believe you'd want https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/fn.eq_expr_value.html here to deny side effects

Comment (llogiq) [reply clippy_lints/src/manual_rotate.rs:87]: @frp yeah, that was my mistake – I totally forgot that we even had this function. We should also add a test case for this, e.g.

```
let l = vec![12_u8, 34];
let y = (l.pop() << 3) + (l.pop() >> 5);
```

which will lint with `eq_expr` but shouldn't with `eq_expr_value`.

Comment (frp) [reply clippy_lints/src/manual_rotate.rs]: Makes perfect sense. Done.

**Comment by llogiq** [comment]
Thank you!

@bors r+

**Review by llogiq — APPROVED** [review]

**Comment by llogiq** [comment]
That is new.

@bors retry

**Pull request merged**